COMPILER.md

```diff
@@ -0,0 +1,286 @@
+Compiler theory of operation
```

Comment:
instead of "COMPILER.md" maybe consider "HOW_COMPILER_WORKS.md" or something.

Reply:
I will consider a better name.